cpp/tensorrt_llm/thop/fuseAGemmOp.cpp

```diff
@@ -0,0 +1,104 @@
+ */
+
+#include "tensorrt_llm/common/opUtils.h"
+#include "tensorrt_llm/kernels/fuseAGemm.h"
+#include "tensorrt_llm/runtime/torchUtils.h"
+
+namespace th = torch;
+namespace tl = tensorrt_llm;
+namespace tk = tensorrt_llm::kernels;
+
+namespace torch_ext
+{
+
+namespace
+{
+template <int kBegin, int kEnd, int kHdIn, int kHdOut>
+struct LoopUnroller
+{
+    static void unroll(int num_tokens, __nv_bfloat16* output, __nv_bfloat16 const* input, __nv_bfloat16 const* weights,
+        cudaStream_t stream)
+    {
+        if (num_tokens == kBegin)
+        {
+            tk::invokeFuseAGemm<__nv_bfloat16, kBegin, kHdIn, kHdOut>(output, input, weights, stream);
+        }
+        else
+        {
+            LoopUnroller<kBegin + 1, kEnd, kHdIn, kHdOut>::unroll(num_tokens, output, input, weights, stream);
+        }
+    }
+};
+
+template <int kEnd, int kHdIn, int kHdOut>
+struct LoopUnroller<kEnd, kEnd, kHdIn, kHdOut>
+{
+    static void unroll(int num_tokens, __nv_bfloat16* output, __nv_bfloat16 const* input, __nv_bfloat16 const* weights,
+        cudaStream_t stream)
+    {
+        if (num_tokens == kEnd)
+        {
+            tk::invokeFuseAGemm<__nv_bfloat16, kEnd, kHdIn, kHdOut>(output, input, weights, stream);
+        }
+        else
+        {
+            throw std::invalid_argument("Invalid num_tokens, only supports 1 to 16");
+        }
+    }
+};
+} // namespace
+
+th::Tensor fuse_a_gemm_op(th::Tensor const& mat_a, th::Tensor const& mat_b, std::optional<at::Tensor> const& bias,
+    std::optional<c10::ScalarType> const& out_dtype)
+{
+    int const num_tokens = mat_a.size(0);
+    int const hd_in = mat_a.size(1);
+    int const hd_out = mat_b.size(1);
+    // auto const out_dtype_ = out_dtype.value_or(mat_a.scalar_type());
+    auto const out_dtype_ = out_dtype.value_or(mat_a.scalar_type());
+    auto const data_type = mat_a.scalar_type();
+    constexpr int kHdIn = 7168;
+    constexpr int kHdOut = 2112;
+    TORCH_CHECK(data_type == at::kBFloat16, "input tensor must be bfloat16");
+    TORCH_CHECK(out_dtype_ == at::kBFloat16, "output tensor must be bfloat16");
+    TORCH_CHECK(hd_in == kHdIn, "hd_in must be 7168");
+    TORCH_CHECK(hd_out == kHdOut, "hd_out must be 2112");
+    TORCH_CHECK(num_tokens >= 1 && num_tokens <= 16, "num_tokens must be between 1 and 16");
+    std::vector<int64_t> output_size = {num_tokens, hd_out};
+    th::Tensor out = th::empty(output_size, mat_a.options().dtype(out_dtype_));
+
+    auto stream = at::cuda::getCurrentCUDAStream(mat_a.get_device());
+
+    LoopUnroller<1, 16, kHdIn, kHdOut>::unroll(num_tokens, reinterpret_cast<__nv_bfloat16*>(out.mutable_data_ptr()),
+        reinterpret_cast<__nv_bfloat16 const*>(mat_a.data_ptr()),
+        reinterpret_cast<__nv_bfloat16 const*>(mat_b.data_ptr()), stream);
+
+    return out;
+}
+
+} // namespace torch_ext
+
+TORCH_LIBRARY_FRAGMENT(trtllm, m)
+{
+    m.def("fuse_a_gemm_op(Tensor mat_a, Tensor mat_b, Tensor? bias, ScalarType? out_dtype) -> (Tensor out)");
```

Comment:
Should we create another directory inside `thop` for all the min-latency kernels for better organization?